File: ts/problems/35-search-insert-position.test.ts
import { describe, expect, test } from "vitest";
import { searchInsert } from "./35-search-insert-position";

describe("searchInsert", () => {
  test.each([
    {
      nums: [],
      target: 0,
      want: 0,
    },
    {
      nums: [1, 3, 5, 6],
      target: 5,
      want: 2,
    },
    {
      nums: [1, 3, 5, 6],
      target: 2,
      want: 1,
    },
    {
      nums: [1, 3, 5, 6],
      target: 7,
      want: 4,
    },
  ])(
    "for $nums with a target of $target, the search position should be $want",
    ({ nums, target, want }) => {
      expect(searchInsert(nums, target)).toBe(want);
    },
  );
});

File: go/problems/35_search_insert_position_test.go
package main

import (
	"fmt"
	"testing"

	"github.com/stretchr/testify/assert"
)

func TestSearchInsert(t *testing.T) {
	tests := []struct {
		nums   []int
		target int
		want   int
	}{
		{
			nums:   []int{},
			target: 0,
			want:   0,
		},
		{
			nums:   []int{1, 3, 5, 6},
			target: 5,
			want:   2,
		},
		{
			nums:   []int{1, 3, 5, 6},
			target: 2,
			want:   1,
		},
		{
			nums:   []int{1, 3, 5, 6},
			target: 7,
			want:   4,
		},
	}

	for _, tt := range tests {
		t.Run(fmt.Sprintf("for %v with a target of %d, the insert position should be %d", tt.nums, tt.target, tt.want), func(t *testing.T) {
			got := searchInsert(tt.nums, tt.target)
			assert.Equal(t, tt.want, got, "got %d want %d", got, tt.want)
		})
	}
}
